node/core/av-store/src/lib.rs

```diff
@@ -477,56 +474,69 @@ fn get_next_chunk_pruning_time(db: &Arc<dyn KeyValueDB>) -> Option<NextChunkPrun
 }
 
 async fn run<Context>(mut subsystem: AvailabilityStoreSubsystem, mut ctx: Context)
-	-> Result<(), Error>
 where
 	Context: SubsystemContext<Message=AvailabilityStoreMessage>,
 {
 	loop {
-		// Every time the following two methods are called a read from DB is performed.
-		// But given that these are very small values which are essentially a newtype
-		// wrappers around `Duration` (`NextChunkPruning` and `NextPoVPruning`) and also the
-		// fact of the frequent reads itself we assume these to end up cached in the memory
-		// anyway and thus these db reads to be reasonably fast.
-	 	let pov_pruning_time = subsystem.maybe_prune_povs()?;
-		let chunk_pruning_time = subsystem.maybe_prune_chunks()?;
-
-		let mut pov_pruning_time = pov_pruning_time.fuse();
-		let mut chunk_pruning_time = chunk_pruning_time.fuse();
-
-		select! {
-			incoming = ctx.recv().fuse() => {
-				match incoming {
-					Ok(FromOverseer::Signal(OverseerSignal::Conclude)) => break,
-					Ok(FromOverseer::Signal(OverseerSignal::ActiveLeaves(
-						ActiveLeavesUpdate { activated, .. })
-					)) => {
-						for activated in activated.into_iter() {
-							process_block_activated(&mut ctx, &subsystem.inner, activated).await?;
-						}
-					}
-					Ok(FromOverseer::Signal(OverseerSignal::BlockFinalized(hash))) => {
-						process_block_finalized(&subsystem, &mut ctx, &subsystem.inner, hash).await?;
-					}
-					Ok(FromOverseer::Communication { msg }) => {
-						process_message(&mut subsystem, &mut ctx, msg).await?;
+		let res = run_iteration(&mut subsystem, &mut ctx).await;
+		match res {
+			Err(e) => {
+				log::log!(target: LOG_TARGET, e.severity(), "{}", e);
+			}
+			Ok(true) => {
+				log::info!(target: LOG_TARGET, "received `Conclude` signal, exiting");
+				break;
+			},
+			Ok(false) => continue,
+		}
+	}
+}
+
+async fn run_iteration<Context>(subsystem: &mut AvailabilityStoreSubsystem, ctx: &mut Context)
```

Comment:
It doesn't apply to this subsystem, but for other subsystems we need to be careful that `run_iteration` has the entire state of the subsystem passed into it. The more functional subsystems might require a few parameters or a refactoring to a larger `State` struct.